### DISCUSSION TITLE
SQLite database to PostgreSQL conversion

### DISCUSSION THREAD

**Discussion by admincheg** [btcpayserver/btcpayserver#4662] [Technical Support]
This information can be useful for somebody in future and I leave it here.

We stuck on very old version of BTCPAY like 1.0.14.3 and used hapinesly before nbxplorer completely crashed. After investigation we decided migration to new version, but our migrations was failed because --sqlite and --postgres at the same time did not works. (It was an error like "Table 'Settings' is not exists", but if you apply migrates to postgresql before running migration from sqlite, you'll got another error - "migration supports only empty database", but anyway)

After that we decided to use fresh one installation because all our payment system was stuck, but history still was interested for us and below I tell you about my adventure.

Some dates in some tables was as '2020-12-18 10:11:32+00:00' and '1305050221139195904' at the same time and it was the main problem of history migration. Few hours in the Internet and experience brought me to source code of .Net and beautiful article:

https://nitratine.net/blog/post/a-warning-for-ef-cores-datetimeoffsettobinaryconverter/

Thanks very much to that guy. It was a big boost in my investigation and after some additional works I've made SQL script for data normalization:

```
--- Create Invoices migration table
CREATE TABLE "Invoices_Fxd" (
    "Id" TEXT NOT NULL CONSTRAINT "PK_Invoices_Fxd" PRIMARY KEY,
    "Blob" BLOB NULL,
    "Created" TEXT NOT NULL,
    "CustomerEmail" TEXT NULL,
    "ExceptionStatus" TEXT NULL,
    "ItemCode" TEXT NULL,
    "OrderId" TEXT NULL,
    "Status" TEXT NULL,
    "StoreDataId" TEXT NULL
);

--- Migrate strange Microsoft binary DateTime to normal format
INSERT INTO Invoices_Fxd
SELECT  
	Id,
	Blob,
	DATETIME((CAST(Created AS INT) >> 11) / 10000 - 62135596800, 'unixepoch') AS Created,
	CustomerEmail,
	ExceptionStatus,
	ItemCode,
	OrderId,
	Status,
	StoreDataId
FROM
	Invoices 
WHERE
	"Created" LIKE '13%';

--- Migrate DateTime format with timezone to simple
SELECT 
	Id,
	Blob,
	substr(Created, 1, 19) AS Created,
	CustomerEmail,
	ExceptionStatus,
	ItemCode,
	OrderId,
	Status,
	StoreDataId
FROM
	Invoices
WHERE
	Created LIKE '20%'
ORDER BY Created DESC;

--- Create temporary migration table
CREATE TABLE "InvoiceEvents_Fxd" (
    "InvoiceDataId" TEXT NOT NULL,
    "UniqueId" TEXT NOT NULL,
    "Message" TEXT NULL,
    "Timestamp" TEXT NOT NULL
);

--- Migrate strange Microsoft binary DateTime to normal format
INSERT INTO InvoiceEvents_Fxd 
select
	InvoiceDataId ,
	UniqueId ,
	Message ,
	DATETIME((CAST(Timestamp AS INT) >> 11) / 10000 - 62135596800, 'unixepoch') AS Timestamp
from 
	"InvoiceEvents"
WHERE
	Timestamp LIKE '13%' ;

--- Migrate DateTime format with timezone to simple
INSERT INTO InvoiceEvents_Fxd 
select
	InvoiceDataId ,
	UniqueId ,
	Message ,
	substr(Timestamp, 1, 19) AS Timestamp
from 
	"InvoiceEvents"
WHERE
	Timestamp LIKE '20%' ;
```

After table preparing I used the DBeaver for simple data migration from one database to another (right click on source table, next, choose destination table, fix mappings, next, next, proceed, ????, PROFIT)

I wrote this because I tired to discover topics with topic starters answers like - I fixed that, thank you - and without any explanation.

**Comment by mjlamb** [comment]
Interesting this may be related to the issue I had with migration and @NicolasDorier fixed the migration scripts. 
https://github.com/btcpayserver/btcpayserver/issues/4623
